### ISSUE TITLE
Different result when using multi_index instead of standard indexing

### ISSUE THREAD

**Issue by theilerp** [TileDB-Inc/TileDB-Py#597]
I have successfully been using TileDB-Py on many projects using 3d coordinates and timestamps (gps week seconds) as domains. I came across a project, where the coordinates can be become negative and thus I ended up with "wrong" results using the default indexing structure:

```python
A[starttime:endtime, X0:X1, Y0:Y1, Z0:Z1]
```

The above query returns an empty dict as soon as e.g. X0 and X1 are both negative and raises a DimensionError if e.g. X0 is negative and X1 is positive (lower bound is bigger than upper bound). The problem of negative indexes became clear after reading some other issues and the documentation. I have than found the option of multi_index and was very happy about the results until I ran my unit tests, because some queries now return different results or rather empty dicts even though the slicing should return something.

```python
A[0.0:0.000002, :, :, :] # standard indexing returns a dict with 3 values
A.multi_index[0.0:0.000002, :, :, :] # mult_index returns an empty dict
```

I have debugged the code of multi_index and found that the number of ranges to query is correctly derived (only 1 query required), but as said, the result is empty.

Is this a bug or am I using multi_index in a wrong matter. Any help would be appreciated.

Operating System: Windows
TileDB-Py-Version: 0.8.11
TlieDB-Version: 2.2.9

**Comment by ihnorton** [comment]
Hi @theilerp, thanks for opening the issue. The `A[]` indexer emulates NumPy indexing semantics, so it works best with zero-indexed, unsigned int dimensions (wrap-around semantics don't really make sense for domains with negative endpoint, for example). For other cases, using `multi_index` is encouraged. But there are some differences: the most significant is that `multi_index` ranges are endpoint-inclusive, like TileDB core (unlike Python/NumPy indexing).

If you are able to share the array schema (possibly making the names generic if necessary), and the result coordinates from the first case (`A[0.0:0.000002, :, :, :]`), it would be very helpful for me to debug. (an example array would be great - but I understand if that might not be possible)

**Comment by theilerp** [comment]
Thanks for the quick response, @ihnorton. 

As often, the error became clear when i tried to do a minimal testing example. The problem was, that my schema had more dimensions than I was indexing with, i.e.,, I queried for timestamp (0.0:0.000002) and ignored the 3d coordinates (:,:,:), but in my schema there was an additional dimension - for simplicity I had removed this dimension in the original question. When generating the minimal example with reduced number of dimensions, it did work and finally I found the error in my code.

Interestingly, passing too few slices / indexes doesn't seem to influence the default indexing, but the multi_index cannot handle this. If this behaviour is already known, maybe one could add it to the documentation. In any case the issue can be closed. Sorry for the inconvenience ...

**Comment by ihnorton** [comment]
> Interestingly, passing too few slices / indexes doesn't seem to influence the default indexing, but the multi_index cannot handle this.

Yes, unfortunately this is a known bug - we are working on it and will have a fix out soon -- the intended semantics are that `A.multi_index[0.0:0.000002]` returns all matches for `(0.0, 0.000002)` on the first dimension with no constraints on the remaining dimensions.

I'm glad it is working for now with the work-around, and I will update this issue/close when the indexing issue is resolved.

**Comment by ihnorton** [comment]
> Interestingly, passing too few slices / indexes doesn't seem to influence the default indexing, but the multi_index cannot handle this. 

This issue was fixed in libtiledb 2.4 (tiledb-py 0.10.*). Please let us know if not and we are happy to reopen and investigate further.
